### PR TITLE
Switch to use tox to conveniently check coverage and code style

### PR DESCRIPTION
**Pull request by maxzheng**
Other changes:
1. Updated gitignore
2. Added context managers in_dir/in_temp_dir to ensure compressed files are created in temp dir instead of cwd (ducktape repo)
3. Fix use of bare except as that is not necessary and may cause issues.

This also paves the way for adding Python 3.7 support later.

### CODE REVIEW

**Comment by maxzheng** [comment]
```
[ducktape]$ test
============================================================= test session starts =============================================================
platform linux2 -- Python 2.7.15, pytest-3.6.3, py-1.5.4, pluggy-0.6.0
rootdir: /home/mzheng/workspace/ducktape, inifile: setup.cfg
plugins: xdist-1.22.4, forked-0.2, cov-2.5.1
collected 126 items                                                                                                                           

tests/cluster/check_cluster.py ..                                                                                                       [  1%]
...
tests/utils/check_util.py ...                                                                                                           [100%]

---------- coverage: platform linux2, python 2.7.15-final-0 ----------
Name                                        Stmts   Miss  Cover
---------------------------------------------------------------
ducktape/__init__.py                            1      0   100%
ducktape/__main__.py                            3      3     0%
ducktape/cluster/__init__.py                    3      0   100%
ducktape/cluster/cluster.py                    34      6    82%
ducktape/cluster/cluster_spec.py               47      1    98%
ducktape/cluster/finite_subcluster.py          19      0   100%
ducktape/cluster/json.py                       57      5    91%
ducktape/cluster/linux_remoteaccount.py        17      7    59%
ducktape/cluster/localhost.py                  25      0   100%
ducktape/cluster/node_container.py             64      2    97%
ducktape/cluster/remoteaccount.py             339    217    36%
ducktape/cluster/vagrant.py                    53      7    87%
ducktape/cluster/windows_remoteaccount.py      60     44    27%
ducktape/command_line/__init__.py               0      0   100%
ducktape/command_line/defaults.py              15      0   100%
ducktape/command_line/main.py                 103     58    44%
ducktape/command_line/parse_args.py            56      1    98%
ducktape/errors.py                              4      0   100%
ducktape/json_serializable.py                   6      1    83%
ducktape/mark/__init__.py                       1      0   100%
ducktape/mark/_mark.py                        172     19    89%
ducktape/mark/mark_expander.py                 16      0   100%
ducktape/mark/resource.py                      19      0   100%
ducktape/services/__init__.py                   0      0   100%
ducktape/services/background_thread.py         51      8    84%
ducktape/services/service.py                  163     37    77%
ducktape/services/service_registry.py          61     16    74%
ducktape/template.py                           37      3    92%
ducktape/tests/__init__.py                      0      0   100%
ducktape/tests/event.py                        59      8    86%
ducktape/tests/loader.py                      172      8    95%
ducktape/tests/loggermaker.py                  20      1    95%
ducktape/tests/reporter.py                    116      3    97%
ducktape/tests/result.py                       98     12    88%
ducktape/tests/runner.py                      173     27    84%
ducktape/tests/runner_client.py               166     16    90%
ducktape/tests/scheduler.py                    30      2    93%
ducktape/tests/serde.py                        10      2    80%
ducktape/tests/session.py                      72     46    36%
ducktape/tests/status.py                       12      0   100%
ducktape/tests/test.py                        237     43    82%
ducktape/utils/__init__.py                      0      0   100%
ducktape/utils/http_utils.py                   14      2    86%
ducktape/utils/local_filesystem_utils.py        9      1    89%
ducktape/utils/terminal_size.py                60     28    53%
ducktape/utils/util.py                         20      2    90%
---------------------------------------------------------------
TOTAL                                        2694    636    76%
Coverage annotated source written to dir textcov
Coverage HTML written to dir htmlcov
Coverage XML written to file coverage.xml

Required test coverage of 70% reached. Total coverage: 76.39%

========================================================= 126 passed in 12.92 seconds =========================================================
cover: OK
style: OK
```

File: .gitignore [hunk 1]
```diff
@@ -1,16 +1,105 @@
...
+coverage.xml
+*.cover
+.hypothesis/
+.pytest_cache/
+
+# Translations
+*.mo
+*.pot
+
+# Django stuff:
```
Comment (ewencp) [inline .gitignore]: i don't care about much stuff that's super-generic for python, but we have no django, jupyter, celery, sage, etc stuff. we should add stuff to `.gitignore` when we have good reason, not if there's an off-chance we might hit it...

File: .gitignore [hunk 2]
```diff
@@ -1,16 +1,105 @@
...
+ENV/
+env.bak/
+venv.bak/
+
+# Spyder project settings
+.spyderproject
+.spyproject
+
+# Rope project settings
+.ropeproject
```
Comment (ewencp) [inline .gitignore]: probably this and files above it too. i don't even know what these files do.

File: .gitignore (hunk 1 above)
Comment (maxzheng) [reply .gitignore]: It's just the latest template from GitHub. We can choose and pick, but that's just more work / easier to use the latest template to cover the new stuff from tox and coverage reports, etcs.

**Pull request merged**